Scripts/Runtime/Logging/.CSProject/UnityLogListener.cs

```diff
@@ -145,7 +162,10 @@ private static void Init()
 
         public UnityLogListener()
         {
-            Debug.Assert(s_Instance == null, $"There can only be one instance of the {nameof(UnityLogListener)} and it has already been initialized.");
+            if (s_Instance != null)
+            {
+                throw new Exception($"There can only be one instance of the {nameof(UnityLogListener)} and it has already been initialized.");
+            }
```

Comment:
I realized that all of the `Debug.Assert` checks in both the C# or Unity logging libraries are useless, because the DLL's are not debug builds.

In this specific case, I changed the assert to an exception because it's still needed in release builds too (in case an Anvil user tries to make their own listener), but how should the rest of the asserts be handled? Should the logging DLL's be changed to debug builds in the repos? Am I just too lazy, and I should've been switching to debug builds when making changes to those libraries? 😛

Reply:
I'm not sure, I'll leave this one for @mbaker3 to chime in on.

Reply:
uhhh, let me investigate this one a bit. I think there is some compiler magic we can use with Unity to handle this.
With a normal C# project you'd include the project (thinking the `anvil-csharp-core` part) in your solution and be able to easily toggle flags and recompile through there.

Reply:
Check out my proposal at https://github.com/decline-cookies/anvil-unity-core/commit/156e566a8252149c1e044358876c3fd6037e953c

Summary:
 - Publish two DLLs one for DEBUG builds and one for Release builds
    - The debug variant has a `*-debug` suffix in the file name
 - Have Unity auto select them using the [Define Constraints](https://docs.unity3d.com/Manual/ScriptCompilationAssemblyDefinitionFiles.html#conditional-assembly) feature
 - I've gone ahead and assumed that you only want `UNITY_ASSERTIONS` enabled when `DEBUG` is also enabled. The Unity player settings allow assertions to be enabled in release builds but 99% of the time that's not what you want. It's not worth creating a permutation for every possible define combination.

The downside to this approach is that you have to run two different build configs manually anytime you change the logging dll.  For now we can just manually enforce that if one DLL changes the other DLL changes as well. In the future, if this approach is accepted and after #72 we can have the DLLs generated and committed when PRs are merged/posted automatically.

If you guys like this then I can take a similar approach with the `anvil-csharp-core` logging dll

Thoughts?

Reply:
I'm on board with this solution - the assumptions are, the logging DLLs are a necessary evil to satisfy Unity, so we'll treat them as regular parts of Anvil, and some day they'll hopefully be built for us.

I'm not sure CI is even the fix for that? It's more like, when I make a change to a logging system file, I wish it would auto-compile the DLL right there in Unity. But at least external builds would ensure they stay up to date in the repo.

Reply:
I ran into a problem with Define Constraints - Unity is telling me that `!DEBUG` is invalid. You can definitely use `!` to invert conditions, and use any built-in or custom define symbol... except `!DEBUG`, then it gives an invalid condition error and removes it.

Beyond that, there were issues with adding a second DLL with a different assembly name (anvil-unity-logging-debug) which throws off the helpers looking for an assembly called "anvil-unity-logging". I tried instead keeping the assembly name the same in both and simply renaming one with "-debug" after building, but then Unity refuses to load a DLL who's name doesn't match its assembly 😠

Maybe we just leave the DLLs as release builds in the repo, so no one accidentally makes a release build with debug libraries, and during development you just have to make local debug builds yourself?

Reply:
I will defer to your and @mbaker3 's judgement on this as I'm not super well versed. 

If `!DEBUG` doesn't work, are you saying that something like `!ANVIL_DEBUG` would work? 

Because then maybe we could do something like this in Anvil core code? 

```csharp
#ifdef DEBUG
#define ANVIL_DEBUG
#endif
```

Although I think what I just wrote is C/C++ specific and you'd have to make it an Editor script that modified the player settings. 

Anyway just food for thought. 

I think ultimately we just don't want to have to do anything annoying or manual just regular development because it won't get done. If it's an annoying/manual one time thing when making changes to the logging projects/dll I think that's fine since it's pretty infrequent.

Reply:
> I'm not sure CI is even the fix for that? It's more like, when I make a change to a logging system file, I wish it would auto-compile the DLL right there in Unity. But at least external builds would ensure they stay up to date in the repo.

There isn't really a way to keep the DLLs up to date. It could always just be a check to make sure the checksum of the DLLs match what you would get if you compiled the source proposed in the merge. That would catch the same issue. A problem for another day.

> I ran into a problem with Define Constraints - Unity is telling me that !DEBUG is invalid. You can definitely use ! to invert conditions, and use any built-in or custom define symbol... except !DEBUG, then it gives an invalid condition error and removes it.

I didn't have this issue. I did notice you had to re-select the DLL to get the inspector to refresh after applying the constraint. I also saw a red ❗ in the inspector but that seemed to indicate that the condition was not met vs a ✔️ that was shown when the condition was met on the other DLL.
For context I was using Unity 2021.3.5. Maybe there are issues with other versions?

> Beyond that, there were issues with adding a second DLL with a different assembly name (anvil-unity-logging-debug) which throws off the helpers looking for an assembly called "anvil-unity-logging". I tried instead keeping the assembly name the same in both and simply renaming one with "-debug" after building, but then Unity refuses to load a DLL who's name doesn't match its assembly 😠

Not sure I understand this one. Could the helpers not just conditionally look for "-debug" in the name when `#if DEBUG`?

Reply:
> I also saw a red ❗ in the inspector but that seemed to indicate that the condition was not met vs a ✔️ that was shown when the condition was met on the other DLL.

If you hover over the ❗ it says invalid condition - if you type anything else like `!UNITY_ASSERTIONS` or `!MY_SYMBOL` or `!DEBU` it shows the ✔️ but the second you type `!DEBUG` it changes to the ❗ and when you hit [Apply], it just removes it from the list.

> Could the helpers not just conditionally look for "-debug" in the name when #if DEBUG?

Sure they could, but I didn't want to keep adding layers of future "what was this for?" weirdness - I can keep trying to get it working if you prefer having both debug and release DLLs, but at this point I would vote to keep the implementation simpler and just have release DLLs in the repo.

Reply:
I'm not seeing `!DEBUG` remove on apply. See screenshot. The ❗ does seem to mean the condition isn't met. The reason you're seeing the ❗ with `!DEBUG` is because that condition isn't met. Trying putting in another condition that is clearly false in the current context. You'll get the same visual treatment.
![image](https://user-images.githubusercontent.com/1426795/194129631-5c7072df-82ac-49de-a77a-c6abd0e795fc.png)
![image](https://user-images.githubusercontent.com/1426795/194129706-ad552a67-0223-4023-ad32-e5feafb3b97d.png)

> Sure they could, but I didn't want to keep adding layers of future "what was this for?" weirdness

I understand now. It's `LogProjectUtils` that's having trouble auto updating references in the csproj automatically.

> ...I would vote to keep the implementation simpler and just have release DLLs in the repo.

yea, this has drifted out of scope. I think it's worth investigating some other day. Will create a reference issue.

Reply:
You're right - when you hover the ❗ the message is "Define constraint is incompatible" (and the one at the top of the list is "One or more define constraints is invalid or incompatible") which I took to mean "this is wrong, fix it" but apparently means the condition is just false right now 😆

I can't get it to remove `!DEBUG` again, maybe it reimported the DLL once or something before, but yeah it does appear to be working as expected. Happy to leave this to another task/day though.